flips/20220715-cadence-purity-analysis.md

```diff
@@ -0,0 +1,292 @@
+but if we could enforce immutability on these fields statically, we could prevent these kinds of attacks without having to ban
+public fields entirely. 
+
+An [issue](https://github.com/onflow/cadence/issues/1805) with function pre- and post-conditions allows arbitrary functions to be
+called in these condition blocks, which can allow implementers to write underhanded code by modifying state inside of a condition. We 
+could solve this by banning all function calls inside condition blocks, but a less restrictive solution could make use of function
+purity information to only ban calls to "impure" functions in conditions. 
+
+This analysis would enable more ergonomic solutions to all of the above problems, as well as generally making it easier for developers
+to see at a glance where state modification (and side-effects in general) are occuring in their code. 
+
+## Design Proposal
+
+### Parsing
+
+This adds two new keywords to the langauge: `pure` and `impure`. These are optionally placed before the `fun` keyword, but after the access modifier, 
+in function declarations and expressions. So, for example, all of the following are valid:
+
+```cadence
+pub pure fun foo(): Void {}
+```
+
+```cadence
+impure fun foo(): Void {}
+```
+
+```cadence
+let x = pure fun(): Void {}
+```
+
+```cadence
+pub struct S {
+    pub pure fun foo(): Void {}
+    impure init()
+}
+```
+
+```cadence
+pub resource interface R {
+    priv pure fun foo(): Void {}
+    impure init()
+}
+```
+
+It is syntactically valid to omit the purity modifier in a function declaration or expression, although in circumstances the type checker
+may reject code that is lacking required annotations (see below).
+
+Function types can also have purity annotations, to be placed after the opening parenthesis but before the parameter list. So, for example, these are valid
+types:
+
+```cadence
+    let f: (pure (Int): Int) = ...
+    let g: (impure (String): String) = ...
+    let h: (pure (): (impure (): Void)) = ...
+```
+
+Any function types without an explicit purity annotation will be considered `impure`. 
+
+### Type Checking
+
+A function that is `pure` (e.g. if it was explicitly declared with the `pure` keyword) may not perform any operations with side effects 
+within its body. Operations with side effects for the purpose of this analysis are:
+
+* Calling a function that is not `pure`
+
+* Mutating or modifying a value that was declared outside the scope of the function body (including the parameters themselves). Variables 
+that are entirely local to the function may be modified as normal. So, for example, this code would be allowed:
+
+    ```cadence
+    pure fun foo(): Int {
+        let a: [Int] = []
+        a.append(3)
+        return a.length
+    }
+    ```
+
+    while the following are not:
+
+    ```cadence
+    pure fun foo(a: [Int]): Int {
+        a.append(3)
+        return a.length
+    }
+    ```
```

Comment:
Ah ok, so it seems like we should be able to allow function arguments to have mutations since they are copied. However, the core of this example remains the same if `a` is captured from outside; so I will change the FLIP to demonstrate this instead.